tests/Module/Checkout/RateCalculatorTest.php

```diff
@@ -0,0 +1,251 @@
+				'cartWeightKg'      => 31,
+				'isCouponApplied'   => false,
+			],
+			'free shipping threshold must be reached'     => [
+				'expectedCost'      => 0,
+				CarrierOptionsDummyFactory::getDefaultCarrier(),
+				'totalProductValue' => 15000,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => false,
+			],
+			'free shipping threshold must not be reached' => [
+				'expectedCost'      => 33,
+				CarrierOptionsDummyFactory::getNoFreeShippingLimitCarrier(),
+				'totalProductValue' => 15000,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => false,
+			],
+			'free shipping coupon must affect cost'       => [
+				'expectedCost'      => 0,
+				CarrierOptionsDummyFactory::getDefaultCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => true,
+			],
+			'free shipping coupon must not affect cost'   => [
+				'expectedCost'      => 33,
+				CarrierOptionsDummyFactory::getNoCouponCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => true,
+			],
+		];
+	}
+
+	/**
+	 * @dataProvider calculationDataProvider
+	 */
+	public function testGetShippingRateCost( ?float $expectedCost, Options $carrierOptions, float $totalProductValue, float $cartWeight, bool $isCouponApplied ): void {
+		$this->createRateCalculatorMock();
+
+		$cost = $this->rateCalculator->getShippingRateCost(
+			$carrierOptions,
+			$totalProductValue,
+			$totalProductValue,
+			$cartWeight,
+			$isCouponApplied
+		);
+
+		self::assertEquals( $expectedCost, $cost );
+	}
+
+	public function testGetCODSurchargeDefault(): void {
+		$this->createRateCalculatorMock();
+
+		$carrierOptions = [];
+		$cartPrice      = 100;
+
+		$surcharge = $this->rateCalculator->getCODSurcharge( $carrierOptions, $cartPrice );
+
+		self::assertEquals( 0.0, $surcharge );
+	}
+
+	public function testGetCODSurchargeWithDefault(): void {
+		$this->createRateCalculatorMock();
+
+		$carrierOptions = [ 'default_COD_surcharge' => 15 ];
+		$cartPrice      = 100;
+
+		$surcharge = $this->rateCalculator->getCODSurcharge( $carrierOptions, $cartPrice );
+
+		self::assertEquals( 15.0, $surcharge );
+	}
+
+	public function testGetCODSurchargeWithSurchargeLimits(): void {
+		$this->createRateCalculatorMock();
+
+		$carrierOptions = [
+			'surcharge_limits'      => [
+				[
+					'order_price' => 50,
+					'surcharge'   => 10,
+				],
+				[
+					'order_price' => 150,
+					'surcharge'   => 5,
+				],
+			],
+			'default_COD_surcharge' => 15,
+		];
+		$cartPrice      = 100;
+
+		$surcharge = $this->rateCalculator->getCODSurcharge( $carrierOptions, $cartPrice );
+
+		self::assertEquals( 5.0, $surcharge );
+	}
+
+	public function testGetCODSurchargeBelowSurchargeLimits(): void {
+		$this->createRateCalculatorMock();
+
+		$carrierOptions = [
+			'surcharge_limits'      => [
+				[
+					'order_price' => 50,
+					'surcharge'   => 10,
+				],
+				[
+					'order_price' => 150,
+					'surcharge'   => 5,
```

Comment:
Poslal bych tam i nějaké floaty, protože tam pak za týden udělam nějakou "chytrou" optimalizaci a místo aby to vrátilo 10.456 to chybou bude vracet 10.0